### PR TITLE
Added converted rows filegroup declarative table (merge first)

### PR DESCRIPTION
**Pull request by smartguest**
Addresses
https://github.com/microsoft/azuredatastudio/issues/25209
https://github.com/microsoft/azuredatastudio/issues/25210
https://github.com/microsoft/azuredatastudio/issues/25130

Styling is not finalized, but it uses whatever is known of other declarative tables.
RowsFilegroupTable

### CODE REVIEW

**Comment by smartguest** [comment]
A follow PR for the filestream table will follow after this one

**Comment by corivera** [comment]
@smartguest  Do the Default checkboxes still work properly when toggling different rows? I remember someone else had issues integrating that behavior with the declarative table

File: extensions/mssql/src/objectManagement/ui/databaseDialog.ts [hunk 1]
```diff
@@ -773,78 +773,94 @@ export class DatabaseDialog extends ObjectManagementDialogBase<Database, Databas
...
+					isReadOnly: true,
+					displayName: localizedConstants.FilesText,
+					headerCssStyles: { ...tableHeader, 'text-align': 'left' },
+				},
+				{
+					valueType: azdata.DeclarativeDataType.boolean,
+					width: 80,
+					displayName: localizedConstants.ReadOnlyText,
+					isReadOnly: false,
+					headerCssStyles: { ...tableHeader, 'text-align': 'center' },
```
Comment (corivera) [inline extensions/mssql/src/objectManagement/ui/databaseDialog.ts:799]: Why the different text alignments here? Some have left and others have center

File: extensions/mssql/src/objectManagement/ui/databaseDialog.ts [hunk 2]
```diff
@@ -773,78 +773,94 @@ export class DatabaseDialog extends ObjectManagementDialogBase<Database, Databas
...
-				// Default column
-				if (arg.column === 3) {
-					this.updateFilegroupsDefaultColumnValues(arg, filegroup, FileGroupType.RowsFileGroup);
-				}
-				// Autogrow all files column
-				if (arg.column === 4) {
-					filegroup.autogrowAllFiles = arg.checked;
-				}
+			this.rowsFilegroupsTable.onDataChanged(async (changedData) => {
+				if (changedData.row && changedData.row > -1) {
```
Comment (corivera) [inline extensions/mssql/src/objectManagement/ui/databaseDialog.ts:835]:
```suggestion
				if (changedData.row > -1) {
```
Greater Than always returns false if one of the arguments is undefined

File: extensions/mssql/src/objectManagement/ui/databaseDialog.ts [hunk 3]
```diff
@@ -1043,6 +1069,37 @@ export class DatabaseDialog extends ObjectManagementDialogBase<Database, Databas
...
+		if (table === this.rowsFilegroupsTable) {
+			newRow.type = FileGroupType.RowsFileGroup;
+			newRow.isReadOnly = false;
+			newRow.isDefault = false;
+			newRow.autogrowAllFiles = false
+			this.objectInfo.filegroups?.push(newRow);
+			newData = this.getDeclarativeTableData(FileGroupType.RowsFileGroup);
+		}
+
+		if (newData !== undefined) {
```
Comment (corivera) [inline extensions/mssql/src/objectManagement/ui/databaseDialog.ts:1095]:
```suggestion
		if (newData) {
```

File: extensions/mssql/src/objectManagement/ui/databaseDialog.ts [hunk 4]
```diff
@@ -1155,6 +1240,35 @@ export class DatabaseDialog extends ObjectManagementDialogBase<Database, Databas
...
+	 * @returns Input component
+	 */
+	private getFilegroupNameInputDeclarative(table: azdata.DeclarativeTableComponent, filegroupType: FileGroupType): azdata.InputBoxComponent {
+		return this.createInputBox(async (value) => {
+			if (table.selectedRow > -1) {
+				let fg = null;
+				if (table === this.rowsFilegroupsTable) {
+					fg = this.rowDataFileGroupsTableRows[table.selectedRow];
+				}
+				if (fg !== null && fg.id < 0) {
```
Comment (corivera) [inline extensions/mssql/src/objectManagement/ui/databaseDialog.ts:1256]:
```suggestion
				if (fg?.id < 0) {
```

File: extensions/mssql/src/ui/dialogBase.ts [hunk 5]
```diff
@@ -411,6 +422,51 @@ export abstract class DialogBase<DialogResult> {
...
 
+	protected addButtonsForDeclarativeTable(table: azdata.DeclarativeTableComponent, addbutton: DialogButton, removeButton: DialogButton, editButton: DialogButton = undefined): azdata.FlexContainer {
+		let addButtonComponent: azdata.ButtonComponent;
+		let editButtonComponent: azdata.ButtonComponent;
+		let removeButtonComponent: azdata.ButtonComponent;
+		let buttonComponents: azdata.ButtonComponent[] = [];
+		const updateButtons = (isRemoveEnabled: boolean = undefined) => {
+			this.onFormFieldChange();
+			const tableSelectedRowsLengthCheck = table.selectedRow > -1 && table.selectedRow < table.dataValues.length;
+			if (editButton !== undefined) {
```
Comment (corivera) [inline extensions/mssql/src/ui/dialogBase.ts:433]:
```suggestion
			if (editButton) {
```

File: extensions/mssql/src/ui/dialogBase.ts [hunk 6]
```diff
@@ -411,6 +422,51 @@ export abstract class DialogBase<DialogResult> {
...
+			addButtonComponent.enabled = this.addButtonEnabled(table);
+			removeButtonComponent.enabled = !!isRemoveEnabled && tableSelectedRowsLengthCheck;
+		}
+		addButtonComponent = this.createButton(uiLoc.AddText, addbutton.buttonAriaLabel, async () => {
+			await addbutton.buttonHandler(addButtonComponent);
+			updateButtons();
+		}, addbutton.enabled ?? true);
+		buttonComponents.push(addButtonComponent);
+
+		if (editButton !== undefined) {
```
Comment (corivera) [inline extensions/mssql/src/ui/dialogBase.ts:445]:
```suggestion
		if (editButton) {
```

**Comment by smartguest** [comment]
Gif showing the current state: RowsFilegroupTableDefault

**Comment by corivera** [comment]
I'm still seeing the rows disappear when toggling defaults with multiple file rows.

**Comment by corivera** [comment]
It's only when the new FileGroup has files in it, since that's when the default checkbox becomes enabled

**Comment by smartguest** [comment]
Added a fix by clearing the dataValues right before setting the new data, this appears to alleviate the angular issue that happens when attempting to set the isDefault changes. (Might need to implement this as part of the extHostModelView but for now it will be only in this area)
FixedDefaultCheckbox

**Review by corivera — APPROVED** [review]

File: extensions/mssql/src/objectManagement/ui/databaseDialog.ts [hunk 7]
```diff
@@ -1155,6 +1241,35 @@ export class DatabaseDialog extends ObjectManagementDialogBase<Database, Databas
 		});
 	}
 
+	/**
+	 * Creates input box for filegroup name (declarative)
+	 * @param table declarative table component
+	 * @param filegroupType filegroup type
+	 * @returns Input component
+	 */
+	private getFilegroupNameInputDeclarative(table: azdata.DeclarativeTableComponent, filegroupType: FileGroupType): azdata.InputBoxComponent {
```
Comment (kisantia) [inline extensions/mssql/src/objectManagement/ui/databaseDialog.ts:1250]: It looks like most of the code here is the same as in the function above in `getFilegroupNameInput()`. Will the above function eventually be removed after the other tables are converted? If so, then not worth spending time refactoring for more code reuse.

**Review by kisantia — APPROVED** [review]

File: extensions/mssql/src/objectManagement/ui/databaseDialog.ts (hunk 7 above)
Comment (smartguest) [reply extensions/mssql/src/objectManagement/ui/databaseDialog.ts:1250]: Yes it will be removed after all the other tables are converted

**Pull request merged**